### PR TITLE
IGNITE-2894

### PR DESCRIPTION
**Pull request by NSAmelchev**
- Add support the Externalizable interface in BinaryMarshaller.
- 103(Object) type used for Externalizable objects.
- No queries support by fields of Externalizable objects such as QuerySqlField. Because Externalizable objects are processed now as binary objects similarly Binarylizable objects with a raw writer format.

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/binary/BinaryClassDescriptor.java [hunk 1]
```diff
@@ -184,17 +207,16 @@ else if (useOptMarshaller)
...
 
         if (useOptMarshaller && userType && !U.isIgnite(cls) && !U.isJdk(cls) && !QueryUtils.isGeometryClass(cls)) {
             U.warn(ctx.log(), "Class \"" + cls.getName() + "\" cannot be serialized using " +
-                BinaryMarshaller.class.getSimpleName() + " because it either implements Externalizable interface " +
-                "or have writeObject/readObject methods. " + OptimizedMarshaller.class.getSimpleName() + " will be " +
-                "used instead and class instances will be deserialized on the server. Please ensure that all nodes " +
-                "have this class in classpath. To enable binary serialization either implement " +
-                Binarylizable.class.getSimpleName() + " interface or set explicit serializer using " +
-                "BinaryTypeConfiguration.setSerializer() method.");
+                BinaryMarshaller.class.getSimpleName() + " because it  have writeObject/readObject methods. " +
```
Comment (voipp) [inline modules/core/src/main/java/org/apache/ignite/internal/binary/BinaryClassDescriptor.java]: it has

File: modules/core/src/main/java/org/apache/ignite/internal/binary/BinaryClassDescriptor.java [hunk 2]
```diff
@@ -489,6 +521,14 @@ public boolean useOptimizedMarshaller() {
     }
 
     /**
+     * @return {@code true} if {@link BinaryMarshaller} marshal the Externalizable objects
```
[the remark this answers is not in the document]
Comment (voipp) [reply modules/core/src/main/java/org/apache/ignite/internal/binary/BinaryClassDescriptor.java]: package private -> public

File: modules/core/src/main/java/org/apache/ignite/internal/binary/BinaryUtils.java [hunk 3]
```diff
@@ -2215,6 +2229,74 @@ public static boolean isCustomJavaSerialization(Class cls) {
...
+
+                if (field.getType() == long.class) {
+                    int mod = field.getModifiers();
+
+                    if (Modifier.isStatic(mod) && Modifier.isFinal(mod)) {
+                        field.setAccessible(true);
+
+                        return (short)field.getLong(null);
+                    }
+                }
```
Comment (voipp) [inline modules/core/src/main/java/org/apache/ignite/internal/binary/BinaryUtils.java]: else throw some exception

File: modules/core/src/main/java/org/apache/ignite/internal/binary/BinaryUtils.java [hunk 4]
```diff
@@ -2215,6 +2229,74 @@ public static boolean isCustomJavaSerialization(Class cls) {
...
+     * Computes the serial version UID value for the given class. The code is taken from {@link
+     * ObjectStreamClass#computeDefaultSUID(Class)}.
+     *
+     * @param cls A class.
+     * @param fields Fields.
+     * @return A serial version UID.
+     * @throws IOException If failed.
+     */
+    @SuppressWarnings("ForLoopReplaceableByForEach")
+    static short computeSerialVersionUid(Class cls, List<Field> fields) throws IOException {
```
Comment (voipp) [inline modules/core/src/main/java/org/apache/ignite/internal/binary/BinaryUtils.java]: ObjectStreamClass#getSerialVersionUID() can be used instead. Just set your class to ObjectStreamClass#cl

File: modules/core/src/test/java/org/apache/ignite/internal/binary/BinaryMarshallerSelfTest.java [hunk 5]
```diff
@@ -593,6 +594,159 @@ public void testExternalizableInEnclosing() throws Exception {
     /**
      * @throws Exception If failed.
      */
+    public void testSimpleExternalizable() throws Exception {
+        SimpleExternalizable exp = new SimpleExternalizable("field");
+
+        SimpleExternalizable obj = marshalUnmarshal(exp);
+
+        assertEquals(obj.field, exp.field);
```
Comment (voipp) [inline modules/core/src/test/java/org/apache/ignite/internal/binary/BinaryMarshallerSelfTest.java]: expected value the first, actual - the second

File: modules/core/src/test/java/org/apache/ignite/internal/binary/BinaryMarshallerSelfTest.java [hunk 6]
```diff
@@ -593,6 +594,159 @@ public void testExternalizableInEnclosing() throws Exception {
...
+        BinaryObject po0 = (BinaryObject)ext.field;
+
+        assertTrue(po.hasField("b"));
+        assertTrue(po.hasField("s"));
+        assertTrue(po.hasField("i"));
+        assertTrue(po.hasField("l"));
+        assertTrue(po.hasField("f"));
+        assertTrue(po.hasField("d"));
+        assertTrue(po.hasField("c"));
+        assertTrue(po.hasField("bool"));
```
Comment (voipp) [inline modules/core/src/test/java/org/apache/ignite/internal/binary/BinaryMarshallerSelfTest.java]: we should check all other fields : str, uuid, date, ts, time, bArr etc.

File: modules/core/src/test/java/org/apache/ignite/internal/binary/BinaryMarshallerSelfTest.java [hunk 7]
```diff
@@ -5177,6 +5331,173 @@ public SimpleExternalizable(String field) {
     }
 
     /**
+     *
```
Comment (voipp) [inline modules/core/src/test/java/org/apache/ignite/internal/binary/BinaryMarshallerSelfTest.java]: add meaningful description

File: modules/core/src/test/java/org/apache/ignite/internal/binary/BinaryObjectBuilderAdditionalSelfTest.java [hunk 8]
```diff
@@ -1536,28 +1536,39 @@ public void testBuilderExternalizable() throws Exception {
...
+        assertEquals(exp, ((BinaryObject)extObj.field("extVal")).deserialize());
+        Assert.assertArrayEquals(expArr, deserializeBinaryObjectArray((Object[])extObj.field("extArr")));
         assertEquals(Integer.valueOf(10), extObj.field("intVal"));
         assertEquals("some string", extObj.field("strVal"));
     }
 
     /**
+     * Deserialize each element of binary objects array.
+     */
+    private Object[] deserializeBinaryObjectArray(Object[] arr){
```
Comment (voipp) [inline modules/core/src/test/java/org/apache/ignite/internal/binary/BinaryObjectBuilderAdditionalSelfTest.java]: may be deserialize into separate array and then return it

File: modules/core/src/main/java/org/apache/ignite/internal/binary/BinaryObjectImpl.java [hunk 9]
```diff
@@ -81,7 +81,7 @@
      * For {@link Externalizable}.
      */
     public BinaryObjectImpl() {
-        // No-op.
```
Comment (NSAmelchev) [inline modules/core/src/main/java/org/apache/ignite/internal/binary/BinaryObjectImpl.java]: may be return // No-op.

File: modules/core/src/test/java/org/apache/ignite/internal/binary/BinaryObjectBuilderAdditionalSelfTest.java [hunk 10]
```diff
@@ -1536,28 +1536,42 @@ public void testBuilderExternalizable() throws Exception {
...
         assertEquals("some string", extObj.field("strVal"));
     }
 
     /**
+     * Deserialize each element of binary objects array.
+     *
+     * @param array Array of binary objects.
+     */
+    private Object[] deserializeBinaryObjectArray(Object[] array){
+        Object[] objArr = new Object[array.length];
```
Comment (NSAmelchev) [inline modules/core/src/test/java/org/apache/ignite/internal/binary/BinaryObjectBuilderAdditionalSelfTest.java]: probel

**Pull request closed without merging**